### ISSUE TITLE
Shifting system clock by several seconds breaks physics in Neos

### ISSUE THREAD

**Issue by shadowpanther** [Neos-Metaverse/NeosPublic#2572]
## Describe the bug
While trying to "fix" #2571, I synced system clock while having Neos running. This broke physics engine. Namely, I fell through the floor, after I was teleported back to spawn my physical locomotion was no longer working, my lasers somewhat worked in worldspace, but not at all in userspace, including context menus. Trying to click the Home tab in Dash actually clicked the Quit tab.

## Relevant issues
#2571 

## To Reproduce
1. Make sure your clock is off by several seconds (6 seconds in my case)
2. Launch Neos
3. After Neos loads and is running, sync the clock
4. Observe how physical interactions break

## Expected behavior
Physics doesn't break.

## Log Files
[VIXEN - 2021.7.1.498 - 2021-07-02 03_22_22.log](https://github.com/Neos-Metaverse/NeosPublic/files/6752048/VIXEN.-.2021.7.1.498.-.2021-07-02.03_22_22.log)

## Bug information (please complete the following information):
 - How often does it happen: Always
 - Does the bug persist after restarting Neos? Yes
 - Neos Version: 2021.7.1.498
 - Neos Platform: Windows
 - Link to reproduction World/Item: 
 - Regression: No/unsure

## Reporters:
Shadow Panther#5154

**Comment by Frooxius** [comment]
I'm not sure this is something we'd spend time addressing. The system time is used very widely across many different behaviors, so there's probably a lot of other things that break as well. Changing all places to somehow compensate for this would likely require a lot of effort, for something that's pretty rare.

It might be that the delta time is the main culprit, if it's that then we can fix it up quickly, but it it's more than that, we'd unlikely prioritize this and just recommend shutting down Neos first if you're going to change your system clock.

**Comment by shadowpanther** [comment]
That's fair, but the main reason I was changing system time is the mentioned #2571. If that was not an issue, I wouldn't sync the clock in the first place.

**Comment by mateon1** [comment]
This may be entirely unfixable because of how `DateTime.UtcNow` is used everywhere in .NET to calculate deltas naïvely, see this issue: https://github.com/dotnet/runtime/issues/15207
The correct solution is to use a monotonic timer, but looking at the linked issue shows that support for that basically doesn't exist it .NET.
If you have some central place where you query times and time deltas from, you can probably build a workaround, from something as simple as `now = LargestTime = Max(LargestTime, DateTime.UtcNow)`, to complex behavior that accelerates/decelerates an internal timer trying to keep it in sync with external time.

**Comment by shadowpanther** [comment]
User Nimbus#2215 had Neos behave similar to what I was experiencing recently. While my case was semi-intentional (I synced the system clock to try to fix another issue), their case seemed to happen randomly. Here's the log:
[Nimbus_crash_log_on_2021.7.3.282.log](https://github.com/Neos-Metaverse/NeosPublic/files/6765206/Nimbus_crash_log_on_2021.7.3.282.log)

Quoting them:

> okay so no bug log; but sometimes when I tab out of Neos and walk away to do something a few things happen:
> 1) the avatar i'm wearing is deleted
> 2) I cannot move
> 3) I am teleported back to spawn
> 4) I cannot use the context menu (as the cursor dissapears when hovering over one of the options)
> 5) the dashboard always highlights the "exit" option, then after clicking that, will only allow me to select "discard homes"
> notes about the bug:
> - It has not ever happened when I was tabbed in
> - It has always been fixed by exiting and reopening Neos
> - It seems to take longer than 5 minutes of idling to occur
https://discord.com/channels/402159838827905024/439422161657135106/861619104430948373

I see the same exception that I have in my log:
```
System.ArithmeticException: Function does not accept floating point Not-a-Number values.
  at System.Math.Sign (System.Single value) [0x00028] in <9577ac7a62ef43179789031239ba8798>:0 
  at BEPUphysics.CollisionShapes.ConvexShapes.BoxShape.GetLocalExtremePointWithoutMargin (BEPUutilities.Vector3& direction, BEPUutilities.Vector3& extremePoint) [0x00000] in <1a09b8ea9f0c48dc8594d42cff1bb16a>:0 
```
Right when the described symptoms happen.

I know that BEPU1 would be removed anyways, but maybe this helps to avoid this problem with BEPU2 in the future.

**Comment by Frooxius** [comment]
@mateon1 Like I mentioned in my post above, fixing the deltas is simple on our end since we control those. However main concern is everything else that uses the system time, including various libraries and other code that'd be very tedious to adjust.

@shadowpanther Did they change their system time as well? Or did it shift somehow? The exception you listed can have lots of different causes, is there anything particular that leads you to believe it's the same one as this?

**Comment by shadowpanther** [comment]
They were away from their computer at the moment it happened and Neos was defocused if that makes any difference. They are using Desktop mode.
It might be an entirely different cause, but the exception is the same and the way Neos's physics broke is the same. User was destroyed and respawned, locomotion didn't enable after spawn, raycasts were broken, there was no way to interact with Dash.

**Comment by Frooxius** [comment]
Do they have anything that shifts the system clock in such scenario? The exception is really very generic one (akin to a Null Reference Exception, you'll see that with a lot of bugs too with completely different causes), it just means there's some corruption in the physics engine.

**Comment by shadowpanther** [comment]
The one reason I suspected was pc going to sleep, but Nimbus confirmed that that was **not** the case. I can't think of another cause that might just shift the clock besides OS syncing time in the background. But from their own words, it happened several times already, and from what I know of Windows time sync, it shouldn't happen more often than once a week by default.

**Comment by shadowpanther** [comment]
Now that I have a tool to sync my system time every 12 hours instead of 7 days to mitigate #2571, whenever the clock syncs, Neos's physics breaks. It just happened again, with the clock going half a second backwards.
![image](https://user-images.githubusercontent.com/675104/125688494-309c4235-cd02-41af-92ac-e9a0e6738404.png)
I know this could not be easily fixed, but unless #2571 is fixed I have no other choice than to sync the time more frequently.

**Comment by Frooxius** [comment]
This should be now fixed in 2021.8.14.443! From what I tested it seems to help prevent the world from getting corrupted. There might be other behaviors that get messed up due to external libraries, but there should definitely be an improvement. 

Let me know if you run into any more issues.
